src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs

```diff
@@ -124,18 +125,25 @@ protected static StringValues AppendValue(in StringValues existing, string appen
 
         protected static int BitCount(long value)
```

Comment:
It is probably worthwhile to change this to something like the following:

```
[MethodImpl(MethodImplOptions.AggressiveInlining)]
protected static int BitCount(long value)
{
	if (Popcnt.X64.IsSupported)
	{
		return (int)Popcnt.X64.PopCount((ulong)value);
	}
	
	return SoftwareFallback((ulong)value);

	int SoftwareFallback(ulong v)	
	{
		// see https://github.com/dotnet/corefx/blob/5965fd3756bc9dd9c89a27621eb10c6931126de2/src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/BitArithmetic.cs

		const ulong Mask01010101 = 0x5555555555555555UL;
		const ulong Mask00110011 = 0x3333333333333333UL;
		const ulong Mask00001111 = 0x0F0F0F0F0F0F0F0FUL;
		const ulong Mask00000001 = 0x0101010101010101UL;

		v = v - ((v >> 1) & Mask01010101);
		v = (v & Mask00110011) + ((v >> 2) & Mask00110011);
		return (int)(unchecked(((v + (v >> 4)) & Mask00001111) * Mask00000001) >> 56);
	}
}
```

Most of the methods that contain the `if (Isa.IsSupported)` checks end up being to big to consider for inlining (and the JIT isn't currently smart enough, at the time inlining is decided, to know that most of the code is dead).

Doing the above ensures that you either get a call directly to the software fallback (which is considered for inlining normally) or you get the intrinsic itself directly inlined.

Reply:
Changed, though moved the local function to the top as @jbogard can sense a disturbance when someone puts a local function at the bottom.

Reply:
> call directly to the software fallback (which is considered for inlining normally)

If the local function is too big for inlining, it can't be forced / hinted with `AggressiveInlining` either, as local functions don't allow attributes.

So this pattern may not always be preferable?!

Though considering that the isa is supported, it's a relly nice pattern.